File: src/sources/base/docker_path_translation.py
import os 
from pathlib import Path

LOSEME_HOST_ROOT = os.getenv("LOSEME_HOST_ROOT")
LOSEME_CONTAINER_ROOT = os.getenv("LOSEME_CONTAINER_ROOT")

def host_path_to_container(host_path: str) -> Path:
    """Translate a host absolute path to its container equivalent."""
    if not LOSEME_HOST_ROOT or not LOSEME_CONTAINER_ROOT:
        raise ValueError("LOSEME_HOST_ROOT and LOSEME_CONTAINER_ROOT environment variables must be set")
    
    host_root = Path(LOSEME_HOST_ROOT).resolve()
    container_root = Path(LOSEME_CONTAINER_ROOT).resolve()
    
    host_path = Path(host_path).resolve()
    
    if not host_path.is_absolute():
        raise ValueError(f"Host path must be absolute: {host_path}")
    
    try:
        relative_path = host_path.relative_to(host_root)
    except ValueError:
        raise ValueError(f"Host path {host_path} is not under the configured host root {host_root}")
    
    container_path = container_root / relative_path
    return container_path


def container_path_to_host(container_path: str) -> Path:
    """Translate a container absolute path to its host equivalent."""
    if not LOSEME_HOST_ROOT or not LOSEME_CONTAINER_ROOT:
        raise ValueError("LOSEME_HOST_ROOT and LOSEME_CONTAINER_ROOT environment variables must be set")
    
    host_root = Path(LOSEME_HOST_ROOT).resolve()
    container_root = Path(LOSEME_CONTAINER_ROOT).resolve()
    
    container_path = Path(container_path).resolve()
    
    if not container_path.is_absolute():
        raise ValueError(f"Container path must be absolute: {container_path}")
    
    try:
        relative_path = container_path.relative_to(container_root)
    except ValueError:
        raise ValueError(f"Container path {container_path} is not under the configured container root {container_root}")
    
    host_path = host_root / relative_path
    return host_path

File: client/npm/web-next/lib/docker-path-translation.ts
import path from 'node:path';
import fs from 'node:fs';

/**
 * Ported from core/loseme_core/docker_path_translation.py and
 * client/sources/base/docker_path_translation.py.
 *
 * IMPORTANT: this must only be called from server-side code (Route
 * Handlers / Server Components), never from the browser, since it reads
 * env vars and the local filesystem of whatever device this Next.js
 * process runs on. Mirrors the original constraint that the web client
 * must run on the same machine as the files it previews.
 */

export function isRunningInDocker(): boolean {
  return fs.existsSync('/.dockerenv');
}

function requireRoots(): { hostRoot: string; containerRoot: string } {
  const hostRoot = process.env.LOSEME_HOST_ROOT;
  const containerRoot = process.env.LOSEME_CONTAINER_ROOT;
  if (!hostRoot || !containerRoot) {
    throw new Error(
      'LOSEME_HOST_ROOT and LOSEME_CONTAINER_ROOT environment variables must be set',
    );
  }
  return { hostRoot: path.resolve(hostRoot), containerRoot: path.resolve(containerRoot) };
}

export function hostPathToContainer(hostPath: string): string {
  const { hostRoot, containerRoot } = requireRoots();
  const resolved = path.resolve(hostPath);

  if (!path.isAbsolute(resolved)) {
    throw new Error(`Host path must be absolute: ${resolved}`);
  }
  const relative = path.relative(hostRoot, resolved);
  if (relative.startsWith('..') || path.isAbsolute(relative)) {
    throw new Error(`Host path ${resolved} is not under the configured host root ${hostRoot}`);
  }
  return path.join(containerRoot, relative);
}

export function containerPathToHost(containerPath: string): string {
  const { hostRoot, containerRoot } = requireRoots();
  const resolved = path.resolve(containerPath);

  if (!path.isAbsolute(resolved)) {
    throw new Error(`Container path must be absolute: ${resolved}`);
  }
  const relative = path.relative(containerRoot, resolved);
  if (relative.startsWith('..') || path.isAbsolute(relative)) {
    throw new Error(
      `Container path ${resolved} is not under the configured container root ${containerRoot}`,
    );
  }
  return path.join(hostRoot, relative);
}
